src/service.rs

```diff
@@ -191,7 +186,17 @@ impl Account {
             rest => rest?,
         };
 
-        let main_key = derive_key(&login_challenge.salt, password)?;
+        // Use only first 16 bytes of salt - previous versions generated a 32-byte salt during
+        // signup, but only used the first 16 bytes
+        let salt = login_challenge
+            .salt
+            .get(..SALT_LENGTH)
+            .ok_or(Error::Encryption(
+                "Salt obtained from login challenge too short: expected at least 16 bytes",
+            ))?
+            .try_into()
+            .unwrap();
+        let main_key = derive_key(&salt, password)?;
```

Comment:
I think it's probably fine, though it's harmless to keep it at 32 and much less scary.

Reply:
This isn't actually a change in functionality, the check has just been moved out of `derive_key()` (and now doesn't panic anymore if the received hash is shorter than 16 bytes). I've adjusted the comment though.